rts/motoko-rts/src/stabilization/layout/stable_array.rs

```diff
@@ -0,0 +1,130 @@
+use crate::{
+    memory::{alloc_array, Memory},
+    stabilization::{
+        deserialization::stable_memory_access::StableMemoryAccess,
+        graph_copy::GraphCopy,
+        serialization::{
+            stable_memory_stream::{ScanStream, StableMemoryStream, WriteStream},
+            ArraySlice, SerializationContext,
+        },
+    },
+    types::{size_of, Array, Value, TAG_ARRAY},
+};
+
+use super::{Serializer, StableToSpace, StableValue, StaticScanner};
+
+#[repr(C)]
+pub struct StableArray {
+    array_length: u64,
+    // Dynamically sized body with `array_length` elements, each of `StableValue`.
+}
+
+impl StaticScanner<StableValue> for StableArray {}
+
+impl Serializer<Array> for StableArray {
+    unsafe fn serialize_static_part(
+        _stable_memory: &mut StableMemoryStream,
+        array: *mut Array,
+    ) -> Self {
+        StableArray {
+            array_length: array.len() as u64,
+        }
+    }
+
+    unsafe fn serialize_dynamic_part(
+        stable_memory: &mut StableMemoryStream,
+        main_array: *mut Array,
+    ) {
+        for index in 0..main_array.len() {
+            let main_element = main_array.get(index);
```

Comment:
So this loop has to complete in one increment --- is it possible that large array will cause serialization to fail? Should we maybe worry about  making this incremental too?

Reply:
I believe the only practically critical loop is the scanning of the array, as this triggers Cheney's algorithm to potentially evacuate all the objects referenced in the elements. Therefore, I had to slice that logic to become incremental. The price is however complexity of such self-designed co-routines.

The other loops may take long, but I believe, even for large practical cases, they should still fit in the quite generous DTS-supported instruction limits (40 billion instructions for stabilization increments that are normal update messages, and 200 billion instructions for the actual upgrade).

I will remeasure very large array cases to see if this is practical problem.

Reply:
I tried a 100_000_000-long array and it takes around 8.6 billion instructions to serialize. It is not cheap but probably practically feasible.